File: Day1/main.c
#include <stdio.h>
#include <stdlib.h>

#define BUFSIZE 1024

int calcFuel(int mass) { return mass / 3 - 2; }

int calcTrickleFuel(int mass) {
    int fuel = 0;

    do {
        mass = calcFuel(mass);
        fuel += mass;
    } while (mass > 5);

    return fuel;
}

int main() {
    int totalFuelReq = 0;

    FILE *file = fopen("input.txt", "r");
    char buffer[BUFSIZE], *end;

    while (fgets(buffer, BUFSIZE, file))
        totalFuelReq += calcTrickleFuel(strtol(buffer, &end, 10));

    fclose(file);

    printf("%d\n", totalFuelReq);

    return 0;
}

File: Day1/main.hs
calcFuel :: Integer -> Integer
calcFuel mass | mass > 5  = mass `div` 3 - 2
              | otherwise = 0

calcTrickleFuel :: Integer -> Integer -> Integer
calcTrickleFuel acc 0    = acc
calcTrickleFuel acc mass = calcTrickleFuel (acc+reqFuel) reqFuel
                           where reqFuel = calcFuel mass

main :: IO ()
main = do
    masses <- readFile "input.txt"
    let parsedMasses = map read (lines masses) :: [Integer]
    let totalReqFuel = sum $ map (calcTrickleFuel 0) parsedMasses
    print totalReqFuel
